lib/init.php

```diff
@@ -57,3 +57,122 @@ function gutenberg_menu() {
 	);
 }
 add_action( 'admin_menu', 'gutenberg_menu', 9 );
+
+if ( ! function_exists( 'html_contains_block' ) ) {
+	/**
+	 * Recursively find a block by attribute.
+	 *
+	 * @param  string  $html The html to search in.
+	 * @param  string  $block_name The block type to search for.
+	 * @param  string  $attribute_name The attribute name to search for.
+	 * @param  string  $attribute_value The attribute value to search for.
```

Comment:
If I were using this I'd be asking a few questions:
 - is the `core/` namespace required or optional?
 - what does it mean to provide an attribute name but not a value?
 - how do I search for a block whose attribute is `null`?

```php
/*
 * …
 *
 * Note that it's not possible to search for an attribute whose value is `null`.
 *
 * @param string  $block_name      Find this block type, with an optional "core/" namespace,
 *                                 e.g. "paragraph", "core/paragraph", "my_plugin/my_block".
 * @param ?string $attribute_name  If provided, the block must also contain this attribute.
 * @param ?mixed  $attribute_value If provided, the given attribute's value must also match this.
 */
```

of course we can find ways of searching for `null`, but I think it's fine not to, and simply to note that it can't be done due to conflating the optional parameter with the definite value `null`

Reply:
Could we consider having an array of attributes => values rather than two separate args? In future folks might want to be even more selective about their block but we're going to be baking in to a three argument signature.

The new signature would be:

`(string $block_name, array $attributes)`.

Reply:
I almost suggested this but on the other hand I feel like this is enough for now and expanding to look for more than one is something that could be addressed in the future, probably with another function.

things in this function are optimized _because_ it's only looking for a single kind of block. if we start allowing multiple search blocks then we would have to start unraveling some of what mitigates the performance issues with this. having a separate function that's costlier and which communicates that is the way out of baking in a simple signature.

wanted to share only because I almost added the same kind of comment you left but didn't, and I wanted to share why I thought it might serve us better as-is with a limited interface.

Reply:
I agree with Dennis on this, from the perspective of keeping this focused is a keystone of our performance mitigating solution, for now. Plus I would really like this method to exist in a non public way, possibly used only by this feature here for now. And this feature does fine with one block one attribute one value.